dstep/main.d

```diff
@@ -172,23 +163,70 @@ int main (string[] args)
     return 0;
 }
 
-void showHelp (Configuration config, GetoptResult)
+void showHelp (Configuration config, GetoptResult getoptResult)
 {
     import std.stdio;
     import std.string;
+    import std.algorithm;
+    import std.range;
+
+    struct Entry
+    {
+        this(string option, string help)
+        {
+            this.option = option;
+            this.help = help;
+        }
+
+        this(Option option)
+        {
+            if (option.optShort && option.optLong)
+                this.option = format("%s, %s", option.optShort, option.optLong);
+            else if (option.optShort)
+                this.option = option.optShort;
+            else
+                this.option = option.optLong;
+
+            this.help = option.help;
+
+            auto beginning = findSplitBefore(this.help, "<");
+
+            if (!beginning[0].empty)
+            {
+                auto placeholder = findSplitAfter(beginning[1], ">");
+
+                if (!placeholder[0].empty)
+                    this.option ~= format(" %s", placeholder[0]);
+            }
+        }
+
+        string option;
+        string help;
+    }
+
+    auto customEntries = [
+        Entry("-o, --output <file>", "Write output to <file>."),
+        Entry("-o, --output <directory>", "Write all the files to <directory>, in case of multiple input files."),
+        Entry("-ObjC, --objective-c", "Treat source input file as Objective-C input.")];
+
+    auto generatedEntries = getoptResult.options
+        .filter!(option => !option.help.empty)
+        .map!(option => Entry(option));
+
+    auto entries = chain(customEntries, generatedEntries);
+
+    auto maxLength = entries.map!(entry => entry.option.length).maxPos.front;
 
     writeln("Usage: dstep [options] <input>");
     writeln("Version: ", strip(config.Version));
     writeln();
     writeln("Options:");
-    writeln("    -o, --output <file>          Write output to <file>.");
-    writeln("    -o, --output <directory>     Write all the files to <directory>, in case of multiple input files.");
-    writeln("    -ObjC, --objective-c         Treat source input file as Objective-C input.");
-    writeln("    -x, --language <language>    Treat subsequent input files as having type <language>.");
-    writeln("    -h, --help                   Show this message and exit.");
-    writeln("    --no-comments                Disable translation of comments.");
-    writeln();
-    writeln("All options that Clang accepts can be used as well.");
-    writeln();
-    writeln("Use the `-h' flag for help.");
+
+    foreach (entry; entries)
+        writeln(format("    %-*s %s", cast(int) maxLength + 1, entry.option, entry.help));
+
+    writeln("");
+    writeln("All options that Clang accepts can be used as well. "
+        "Use the `-h' flag for help. "
+        "To disable boolean options use false, e.g. --comments=false.");
```

Comment:
Since most of this is change anyway, is it better to generate the whole string and call `writeln` just once?